### ISSUE TITLE
abstract `__init__` signature of `abc.ABC` doesn't get type checked

### ISSUE THREAD

**Issue by bluenote10** [python/mypy#15500]
**Bug Report**

If an abstract base class has a `__init__` that is annotated as `@abc.abstractmethod`, the signature should be verified for clients implementing the `__init__`. This does not seem to be the case.

**To Reproduce**

```python
import abc

class Base(abc.ABC):
    @abc.abstractmethod
    def __init__(self, a: int):
        ...

class Client(Base):
    def __init__(self, a: str):
        ...
```

**Expected Behavior**

A type error in the child's `__init__` because it violates the signature. This works for regular methods, i.e., replacing both `__init__` by e.g. `foo` errors with:

```
main.py:10: error: Argument 1 of "foo" is incompatible with supertype "Base"; supertype defines the argument type as "int"  [override]
main.py:10: note: This violates the Liskov substitution principle
main.py:10: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

**Actual Behavior**

`mypy` doesn't detect the type error.

**Your Environment**

- Mypy version used: 1.4.0
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): (nothing, just using the playground)
- Python version used: 3.11

**Comment by erictraut** [comment]
I think this is by design. Mypy treats `__init__` as exempt from LSP checks.

**Comment by bluenote10** [comment]
I'm curious what is the motivation for that or whether that's specified somewhere? On first glance, it seems to prevent quite a few seemingly sensible use cases of the type system, so presumable that most be a strong reason for it.

I actually tried several times to use a `Protocol` to express requirements on `__init__` which probably also failed because of that. Checking [PEP 544](https://peps.python.org/pep-0544/) never made it clear why that isn't possible.

**Comment by AlexWaygood** [comment]
If you applied LSP to constructor and initialiser signatures, literally every Python class with a custom `__init__` or `__new__` signature would violate LSP, since all classes inherit from `object`, and `object.__init__` takes 0 arguments other than `self`.

More generally, though, applying LSP to constructor or initialiser methods would be misunderstanding the purpose of the Liskov Substitution Principle. There's lots of great explanation [here](https://softwareengineering.stackexchange.com/questions/431549/the-liskov-substitution-principle-and-python) which go much more in depth than I'm prepared to right now about why LSP generally only applies to methods available _on constructed objects_, and generally isn't seen as applying to methods that are used _to construct objects themselves_.
